### ISSUE TITLE
6.7 date mappings using Java time formats lose `8` prefix

### ISSUE THREAD

**Issue by droberts195** [elastic/elasticsearch#38567]
Version 6.7 supports using Java time formats for dates if they are prefixed with `8`.

It appears that when such formats are used in mappings the `8` prefix gets lost, and hence the Java time format is interpreted as a Joda time format, leading to exceptions or incorrect parsing.

To reproduce:

* Run Elasticsearch 6.7
* Create an index as follows:

```
curl -H "Content-Type: application/json" -XPUT "localhost:9200/airline_data" -d '{
  "mappings": {
    "_doc": {
      "properties": {
        "@timestamp": {
          "type": "date"
        },
        "airline": {
          "type": "keyword"
        },
        "responsetime": {
          "type": "double"
        },
        "sourcetype": {
          "type": "keyword"
        },
        "time": {
          "type": "date",
          "format": "8yyyy-MM-dd HH:mm:ssXX"
        }
      }
    }
  }
}'
```

* Look in the log file - you will see this exception: `Invalid format: [yyyy-MM-dd HH:mm:ssXX]: Illegal pattern component: XX`
* Look at the stack trace - you will see that the format string is being passed to Joda to interpret
* Get the mappings via http://localhost:9200/airline_data/_mappings?pretty - you will see that the `8` has been lost:

```
{
  "airline_data" : {
    "mappings" : {
      "_doc" : {
        "properties" : {
          "@timestamp" : {
            "type" : "date"
          },
          "airline" : {
            "type" : "keyword"
          },
          "responsetime" : {
            "type" : "double"
          },
          "sourcetype" : {
            "type" : "keyword"
          },
          "time" : {
            "type" : "date",
            "format" : "yyyy-MM-dd HH:mm:ssXX"
          }
        }
      }
    }
  }
}
```

In case this is affected by in-flight PRs, this repro was done from 65f807c.

Unless I'm mistaken I think this is a bad bug that needs to be fixed before 6.7.0 release, otherwise the BWC layer for time formats in 6.7 isn't really doing what it's supposed to.

(For 7.x silently losing the `8` prefix is probably a good thing, as it seamlessly upgrades the mappings.)

**Comment by droberts195** [comment]
The response from the creation request is `{"acknowledged":true,"shards_acknowledged":false,"index":"airline_data"}`, so I suspect that the mappings are reserialised and unserialised when sending to the shards, and it's this reserialisation of the mappings that loses the `8`.

**Comment by spinscale** [comment]
I think I know what happens here: The `DateFieldMapper.toXContent()` method writes out the pattern of the formatter in the `format` field. This pattern will just include a format or the name of the pattern, but never the `8` prefix - which it has to include.

**Comment by spinscale** [comment]
closed by #38712
